api/pkg/auth/google_auth.go

```diff
@@ -0,0 +1,76 @@
+package auth
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+
+	"golang.org/x/oauth2"
+	"golang.org/x/oauth2/google"
+	"google.golang.org/api/idtoken"
+	"google.golang.org/api/option"
+	"google.golang.org/api/option/internaloption"
+	htransport "google.golang.org/api/transport/http"
+)
+
+// JSON key file types.
+const (
+	serviceAccountKey = "service_account"
+)
+
+type credentialsFile struct {
+	Type string `json:"type"`
+}
+
+// idTokenSource is an oauth2.TokenSource that wraps another TokenSource
+// It takes the id_token from TokenSource and passes that on as a bearer token
+type idTokenSource struct {
+	TokenSource oauth2.TokenSource
+}
+
+func (s *idTokenSource) Token() (*oauth2.Token, error) {
+	token, err := s.TokenSource.Token()
+	if err != nil {
+		return nil, err
+	}
+
+	idToken, ok := token.Extra("id_token").(string)
+	if !ok {
+		return nil, fmt.Errorf("token did not contain an id_token")
+	}
+
+	return &oauth2.Token{
+		AccessToken: idToken,
+		TokenType:   "Bearer",
+		Expiry:      token.Expiry,
+	}, nil
+}
+
+// InitGoogleClient is a helper method to be used by CaraML components to initialise a Google Client that appends ID
+// tokens to the headers of all outgoing requests with ID tokens, regardless of the type of credentials used
+func InitGoogleClient(ctx context.Context, audience string) (*http.Client, error) {
+	cred, err := google.FindDefaultCredentials(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	var f credentialsFile
+	if err := json.Unmarshal(cred.JSON, &f); err != nil {
+		return nil, err
+	}
+
+	if f.Type == serviceAccountKey {
+		return idtoken.NewClient(ctx, audience)
+	}
+
+	tokenSource := oauth2.ReuseTokenSource(nil, &idTokenSource{TokenSource: cred.TokenSource})
+
+	var opts []idtoken.ClientOption
+	opts = append(opts, option.WithTokenSource(tokenSource), internaloption.SkipDialSettingsValidation())
```

Comment:
Why do we need to append `internaloption.SkipDialSettingsValidation()` ?

Reply:
Hmm, to be honest I haven't looked into this too deeply, it's something that the Google Client library added when they returned a new HTTP client: https://github.com/googleapis/google-api-go-client/blob/64b6ee4ccbd1b9a52eed972989e9b8117406613b/idtoken/idtoken.go#L66

I'll go look a little more into this and if I can't find much information, I'll just test it without that additional argument and see if it works before removing it 🤔